### PR TITLE
🌱 Upgrade to kind v0.11.0

### LINKED ISSUE #4405: Multiple e2e tests are flaky because of error container is not running
**What steps did you take and what happened:**
[A clear and concise description on how to REPRODUCE the bug.]

According to [testgrid](https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-main&width=20&sort-by-failures=&exclude-non-failed-tests=48) `capi-e2e.When following the Cluster API quick-start [PR-Blocking] Should create a workload cluster` and a few others are failing from time to time.

I looked at the last two occurrences in the capi-quickstart test. In both cases a machine did not come up because `mkdir -p /etc/kubernetes/pki` was failing because the respective container was not running. It was retried for a while but the container didn't come up. I tried to find any other logs but couldn't find anything. Logs from the controllers aggregated and sorted for the affected node of this test: https://gist.github.com/sbueringer/e007c989c158d66dd6d3078f8c904f30 ([ProwJob](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1376713791990927360))

I think right now we don't have the necessary data/logs to find out why this happens. I would propose to gather the logs of the Docker service which is used in those tests (the dind used in the ProwJob). Maybe there's something interesting there. Are there any other Docker / kind / ... logs which we could retrieve?

What I found in the kubekins image we're using:
* /var/log/docker.log

**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

I assume the following test failures are related:
* same container not running logs in the capd manager: (I didn't look at all logs, but they all file basically in the same line of code in the tests)
  * capi-e2e: When testing K8S conformance [Conformance] Should create a workload cluster and run kubetest
    * https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1376575756603232256
    * https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1376468293379952640
  * capi-e2e: When testing KCP upgrade Should successfully upgrade Kubernetes, DNS, kube-proxy, and etcd in a HA cluster
    * https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1376652387720105984
  * capi-e2e: When testing KCP upgrade Should successfully upgrade Kubernetes, DNS, kube-proxy, and etcd in a HA cluster
    * https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1376529700758753280

So tl;dr apart from the MachineRemediation most of our other flaky tests are probably caused by this issue. They are usually failing in the following lines of code:
* controlplane_helpers.go:109 https://github.com/kubernetes-sigs/cluster-api/blob/7478817225e0a75acb6e14fc7b438231578073d2/test/framework/controlplane_helpers.go#L109
* controlplane_helpers.go:146 https://github.com/kubernetes-sigs/cluster-api/blob/7478817225e0a75acb6e14fc7b438231578073d2/test/framework/controlplane_helpers.go#L146
* machinedeployment_helper:120: https://github.com/kubernetes-sigs/cluster-api/blob/7478817225e0a75acb6e14fc7b438231578073d2/test/framework/machinedeployment_helpers.go#L120
* machinepool_helpers.go:85: https://github.com/kubernetes-sigs/cluster-api/blob/7478817225e0a75acb6e14fc7b438231578073d2/test/framework/machinepool_helpers.go#L85

Even if they not all have the same root cause. Fixing this error should fix most of them.

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### LINKED ISSUE #4462: Periodic e2e test for workload cluster upgrade from v1.21 to latest are failing
Looking at https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-main-1-21-latest&width=20 those test are consistently failing

```
When upgrading a workload cluster and testing K8S conformance [Conformance] [K8s-Upgrade]
/home/prow/go/src/sigs.k8s.io/cluster-api/test/e2e/cluster_upgrade_test.go:27
  Should create and upgrade a workload cluster and run kubetest [It]
  /home/prow/go/src/sigs.k8s.io/cluster-api/test/e2e/cluster_upgrade.go:82
  Timed out after 900.003s.
  Expected
      <bool>: false
  to be true
  /home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/deployment_helpers.go:248
  Full Stack Trace
  sigs.k8s.io/cluster-api/test/framework.WaitForDNSUpgrade(0x211d800, 0xc0000440d8, 0x7f6bc8430ee0, 0xc00002a150, 0xc00004221b, 0x5, 0x0, 0x0, 0x0)
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/deployment_helpers.go:248 +0x13c
  sigs.k8s.io/cluster-api/test/framework.UpgradeControlPlaneAndWaitForUpgrade(0x211d800, 0xc0000440d8, 0x2134810, 0xc000ba2200, 0xc00059e340, 0xc00084d800, 0xc00004c05e, 0x22, 0xc000042158, 0x8, ...)
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/controlplane_helpers.go:335 +0xe13
  sigs.k8s.io/cluster-api/test/e2e.ClusterUpgradeConformanceSpec.func2()
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/e2e/cluster_upgrade.go:108 +0xa5b
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc000401380, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/internal/leafnodes/runner.go:113 +0xa3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run(0xc000401380, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/internal/leafnodes/runner.go:64 +0x15c
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run(0xc0003fcc20, 0x20e0f20, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample(0xc000502c30, 0x0, 0x20e0f20, 0xc000054880)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/internal/spec/spec.go:215 +0x72f
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run(0xc000502c30, 0x20e0f20, 0xc000054880)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/internal/spec/spec.go:138 +0xf2
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec(0xc0000e6420, 0xc000502c30, 0x0)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/internal/specrunner/spec_runner.go:200 +0x111
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs(0xc0000e6420, 0x1)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/internal/specrunner/spec_runner.go:170 +0x147
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run(0xc0000e6420, 0xc000044470)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/internal/specrunner/spec_runner.go:66 +0x117
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run(0xc0000fc070, 0x7f6bca897d38, 0xc00050f800, 0x1e80cf7, 0x8, 0xc0003fcb00, 0x2, 0x2, 0x2125e58, 0xc000054880, ...)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/internal/suite/suite.go:79 +0x546
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters(0x20e25e0, 0xc00050f800, 0x1e80cf7, 0x8, 0xc0003fcae0, 0x2, 0x2, 0x2)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/ginkgo_dsl.go:229 +0x218
  github.com/onsi/ginkgo.RunSpecsWithDefaultAndCustomReporters(0x20e25e0, 0xc00050f800, 0x1e80cf7, 0x8, 0xc00011df60, 0x1, 0x1, 0x60708fc4)
  	/home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.15.2/ginkgo_dsl.go:217 +0xad
  sigs.k8s.io/cluster-api/test/e2e.TestE2E(0xc00050f800)
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/e2e/e2e_suite_test.go:86 +0x136
  testing.tRunner(0xc00050f800, 0x1f55150)
  	/usr/local/go/src/testing/testing.go:1194 +0xef
  created by testing.(*T).Run
  	/usr/local/go/src/testing/testing.go:1239 +0x2b3
------------------------------
STEP: Tearing down the management cluster
Summarizing 1 Failure:
[Fail] When upgrading a workload cluster and testing K8S conformance [Conformance] [K8s-Upgrade] [It] Should create and upgrade a workload cluster and run kubetest 
/home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/deployment_helpers.go:248

```

/kind failing-test
@srm09 
@sbueringer

### LINKED ISSUE #4463: Periodic e2e test for workload cluster upgrade from v1.20 to v1.21 are failing
Looking at https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-main-1-20-1-21&width=20 those test are consistently failing

/kind failing-test
@srm09
@sbueringer

### LINKED ISSUE #4503: Upgrade to kind v0.11.0 in our e2e tests
**User Story**

As a developer I would like to use the latest (next) kind version to reduce flakiness in the CAPI e2e tests.

**Detailed Description**

I had some issues running our e2e tests with the latest kind version: https://github.com/kubernetes-sigs/kind/commit/216edda697a37f8bf16651af6c23b7e2bb7ef42f . We want to upgrade asap after the next kind release to get the pipefail fix which will reduce flakiness in our tests (xref: https://github.com/kubernetes-sigs/cluster-api/issues/4405).

Goal of this issues is to figure out if there's anything blocking us from upgrading to the new kind version and if there is, how we can overcome these issues. 

**Anything else you would like to add:**

Let's discuss this further in this google doc: https://docs.google.com/document/d/1zsxMrdQoXT01vkEnK9pnbVidtNfglBYoNMkHhEGqqM4/edit#heading=h.pasf2p8ffbhs

It's a fairly complex situation and it's probably easier to discuss there. Next steps would be to gather data if there are problems in some combinations and how to reproduce them.

/kind feature

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

Upgrades to kind v0.11.0. Main reason for that is to get rid of #4405.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4503
Related to  #4405 (I want to verify the results posts-submit before closing #4405)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -37,6 +37,6 @@ require (
 	k8s.io/kubectl v0.21.0-beta.1
 	k8s.io/utils v0.0.0-20210305010621-2afb4311ab10
 	sigs.k8s.io/controller-runtime v0.9.0-alpha.1
-	sigs.k8s.io/kind v0.9.0
+	sigs.k8s.io/kind v0.10.1-0.20210406115533-b6bc11252265
```
Comment (vincepri) [inline go.mod:40]: Was 0.11 not tagged yet?

Comment (sbueringer) [reply go.mod:40]: Afaik current plan is to release & tag next week. I'm mostly testing this to see if it would fix our failed upgrade test.

Comment (sbueringer) [reply go.mod:40]: But overall we have to decide sooner or later if it makes sense to use latest kind version when we try to test against latest Kubernetes. 

I think based on 1.21 (as an example) discussed here: https://github.com/kubernetes-sigs/kind/issues/2189#issuecomment-816313206 it could make sense to stay on the main-branch of kind to get the latest fixes for the latest Kubernetes releases. I guess otherwise we will always hit issues because the last kind release does not support the last Kubernetes release or the last changes on Kubernetes main.

**Comment by sbueringer** [comment]
looks like cluster upgrade worked now, but I hit the 2h timeout which should not happen. Not sure how to debug this:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/4469/pull-cluster-api-e2e-workload-upgrade-1-21-latest-main/1381836594410950656

/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

**Comment by sbueringer** [comment]
Upgrade tests will be fixed here: https://github.com/kubernetes-sigs/cluster-api/pull/4470

I'll keep this PR for now for PoC with the new kind version, but I'll wait until the kind release the next week (s) as we apparently don't need it to get the tests green (at least in Prow, I'm not sure about local)

**Comment by sbueringer** [comment]
Now it's definitely just a test branch which I don't intend to merge.

**Comment by vincepri** [comment]
@sbueringer @fabriziopandini What's the status for this update?

**Comment by sbueringer** [comment]
> @sbueringer @fabriziopandini What's the status for this update?

Mainly that I was busy with other things the last 1-2 weeks and we're currently figuring out how we want to continue once kind is released. I'll will have time tomorrow to continue with PoCs.

The current state is documented here https://docs.google.com/document/d/1zsxMrdQoXT01vkEnK9pnbVidtNfglBYoNMkHhEGqqM4/edit# (corresponding issue https://github.com/kubernetes-sigs/cluster-api/issues/4503)

tl;dr from the doc is: out of the box the current kind main version won't work with Kubernetes  >= 1.21 (because systemd cgroupDriver is default) without adjustments from our side. But there are a lot of easy workarounds imho (like settings cgroupDriver to `cgroupfs`).

**Comment by vincepri** [comment]
Let's make sure (if we haven't done so already) to open an issue in Kind as well?

We can link back to this PR + the google doc linked above.

**Comment by sbueringer** [comment]
> Let's make sure (if there isn't one) to open an issue in Kind as well?
> 
> We can link back to this PR + the google doc linked above.

Sounds good to me. Let's consider this after I got more data tomorrow, this makes it easier to open a more targeted issue at what we actually want from them. The problem is that kind itself works flawlessly.  Our issue is that the kind images don't work out of the box in our scenario. So I want to make sure we're sure what we ask them for.

**Comment by sbueringer** [comment]
Just in case someone is following this PR. I started a discussion in Slack: https://kubernetes.slack.com/archives/C8TSNPY4T/p1620923791417000

For now further actions will be taken on other issues/PRs. I'll keep the current branch for experimentation with the current kind version. But I'll open a new (clean) one once I have something which should be reviewed/merged.

**Comment by sbueringer** [comment]
Rebased after merge of: https://github.com/kubernetes-sigs/cluster-api/pull/4614
Also upgraded to latest kind commit

/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

**Comment by sbueringer** [comment]
@fabriziopandini @vincepri PR should be ready for review now. 

Let me know if I should rather open a clean new PR for it (for easier traceability as this PR already has a lot of back and forth on it because of my PoCs)


/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm

please also update v1alpha3-to-v1alpha4.md https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md#kind-version

**Comment by sbueringer** [comment]
@CecileRobertMichon updated the doc. I'm not sure if we should do anything about the minimum kind version quick start documentation: https://cluster-api.sigs.k8s.io/user/quick-start.html#install-andor-configure-a-kubernetes-cluster

I'm not sure what it means that we "support" this kind version. There's nothing from our side which breaks compatibility with kind < v0.11 but we're only testing kind v0.11.

/cc @fabriziopandini

**Review by vincepri — APPROVED** [review]
/lgtm

**Comment by sbueringer** [comment]
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main
(suspected flake, if it isn't it's helpful to get more data/logs)

**Comment by sbueringer** [comment]
Ah, just forgot that the 1.21=>latest test is expected to fail at the moment

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve
